diesel_cli/src/main.rs

```diff
@@ -186,6 +190,33 @@ fn convert_absolute_path_to_relative(target_path: &Path, mut current_path: &Path
     result.join(target_path.strip_prefix(current_path).unwrap())
 }
 
+fn run_infer_schema(matches: &ArgMatches) {
+    let database_url = database::database_url(matches);
+    let schema_name = matches.value_of("schema");
+    let tables = matches.values_of("table-name").map(|v| v.collect())
+        .unwrap_or(::std::collections::HashSet::new());
+    let is_whitelist = matches.is_present("whitelist");
+    let is_blacklist = matches.is_present("blacklist");
+    let schema = diesel_infer_schema::infer_schema_for_schema_name(&database_url, schema_name,
+        |table_name, error| {
+            println!("Failed to infer schema for table {}: {}", table_name, error);
+        }, |table_name| {
+            if is_whitelist {
+                return tables.contains(table_name);
+            }
+            if is_blacklist {
+                return !tables.contains(table_name);
+            }
+            true
+        })
```

Comment:
I know why you did it this way, but I'm not too happy with the structure of this. I'll try to think of something.